aexpect/ops.py

```diff
@@ -0,0 +1,601 @@
+# This Python file uses the following encoding: utf-8
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Intra2net AG and aexpect contributors
+
+"""
+A list of frequent operations performed through remote sessions.
+
+Most commands passed to a session are routine invocations of common userland
+tools. Nevertheless some of them require a certain effort to set up because
+they are squeezed through the eye of the needle, the shell session.
+
+The functionality in this module assumes a guest with the ordinary userland
+tools available. Some of it might just work on Windows guests if the proper
+environment (i. e. Cygwin) is present.
+
+This module aims to provide convenient wrappers for butter-and-bread
+functionality on a Linux system. Currently, the implemented functions fall
+under a few categories: ``stat``, ``test``, common file ops, common utilities,
+and system utilities.
+
+stat:
+Allows passing an optional format argument (``-c``). Custom wrappers for
+extracting *atime*, *ctime*, and *mtime*, etc exist.
+
+test:
+Allows for testing existence, permissions and other properties of files.
+
+file ops:
+Simple file operations executed through sessions. Most functions boil down to
+executing simple commands like ls, tar, md5sum, etc. If these operations return
+non-0, a :py:class:`RuntimeError` is raised, containing the commands's error
+message. All functions :py:func:`shlex.quote` their args for better security.
+
+utilities:
+More complex linux utilities for tarball extraction, file hashing, etc.
+
+system:
+System related operations like restarting or stopping services.
+"""
+
+import logging
+import re
+from os.path import join
+from os.path import split as path_split
+from shlex import quote
+
+from aexpect.exceptions import ShellCmdError
+
+# Need this import for sphinx and other documentation to produce links later on
+# from .client import ShellSession
```

Comment:
No, this was left since I wasn't sure what to do so your advice is welcome. The problem is that Sphinx will return errors when building the documentation if not all docstring types are found (here we have some `ShellSession` types in the docstrings). However, if I import the `ShellSession` then LGTM will complain that it is empty import not used within the code. So it is a bit of a stalemate. If you have a recommendation I would follow it, here I simply commented it out to make LGTM happy but I still prefer to keep it for any Sphinx documentation we might build in the near future.